README.md
# Startdde

**Description**:
Startdde is used for launching DDE components and invoking
user's custom applications which compliant with xdg autostart specification.

## Dependencies

### Build dependencies

- cmake
- pkg-config
- golang-go
- [go-dlib](https://github.com/linuxdeepin/go-lib)
- [go-fsnotify](https://github.com/howeyc/fsnotify)
- [dde-dbus-factory](https://github.com/linuxdeepin/dbus-factory)
- [go-gir-generator](https://github.com/linuxdeepin/go-gir-generator)
- [dde-api](https://github.com/linuxdeepin/dde-api)
- [go-x11-client](https://github.com/linuxdeepin/go-x11-client)
- libgnome-keyring
- libxfixes
- libxcursor

### Runtime dependencies

- dde-daemon
- deepin-wm | deepin-metacity
- libgnome-keyring
- libxfixes
- libxcursor

## Installation

### Deepin

Install prerequisites
```
$ sudo apt-get build-dep startdde
```

Build
```
$ GOPATH=/usr/share/gocode make
```

If you have isolated testing build environment (say a docker container), you can install it directly
```
$ sudo make install
```

generate package files and install Startdde with it
```
$ debuild -uc -us ...
$ sudo dpkg -i ../startdde-*deb
```

## Usage

Run Startdde with the command below

```
Usage of /usr/bin/startdde:
  -d=false: debug
```

### Directly run without display manager

```
$ echo "dbus-launch --exit-with-session /usr/bin/startdde" > ~/.xinitrc
$ startx
```

### Run with display manager

1. construct a session desktop in /usr/share/xsessions

```
cat /usr/share/xsessions/deepin.desktop

[Desktop Entry]
Name=Deepin
Comment=Deepin Desktop Environment
Exec=/usr/bin/startdde
```

2. Using DisplayManager like, gdm, kdm or lightdm to startup Startdde

## Getting help

Any usage issues can ask for help via

* [Gitter](https://gitter.im/orgs/linuxdeepin/rooms)
* [IRC channel](https://webchat.freenode.net/?channels=deepin)
* [Forum](https://bbs.deepin.org)
* [WiKi](https://wiki.deepin.org/)

## Getting involved

We encourage you to report issues and contribute changes

* [Contribution guide for developers](https://github.com/linuxdeepin/developer-center/wiki/Contribution-Guidelines-for-Developers-en). (English)
* [开发者代码贡献指南](https://github.com/linuxdeepin/developer-center/wiki/Contribution-Guidelines-for-Developers) (中文)

## License

Startdde is licensed under [GPLv3](LICENSE).


CHANGELOG.md
[3.14.6] 2019-05-14
*   feat: handle deepin-greeter missing qt-theme.ini
*   chore: do not set DEEPIN_WINE_SCALE in ~/.dde_env

[3.14.5] 2019-05-09
*   feat(xsettings): set scale factor for deepin wine apps
*   fix(xsettings): send notification the first time auto set scale

[3.14.4] 2019-04-18
*   feat: set ScaleLogicalDpi in qt-theme.ini
*   feat(xsettings): set Xft.dpi in x resources for wps
*   chore(xsettings): do not set Xft.dpi in RESOURCE_MANAGER
*   fix: kwin is not affected by the choice of dde-wm-chooser

[3.14.3] 2019-04-12
*   fix: do not set recommended scale factor for qt app
*   chore: waiting for kwin after launch it
*   fix: not send SetScaleFactorDone signal

[3.14.2] 2019-04-10
*   chore: update depends for wm

[3.14.1] 2019-04-09
*   feat: set kwin compositing with com.deepin.wm
*   feat(watchdog): add task for kwin

[3.14.0] 2019-04-08
*   feat: set screenScaleFactors to a number
*   feat: set scale factor for qt in a new way
*   chore: update build depends

[3.13.3] 2019-04-02
*   fix(display): rotate the screen for the first time, then reset, but failed
*   feat: add wm_kwin.WMSwitcher for control kwin compositor

[3.13.2] 2019-03-28
*   fix: bamfdaemon.service start request repeated too quickly

[3.13.1] 2019-03-27
*   fix(display): cancel rotation, but the rotation of the touchpad is not restored
*   chore: `auto_launch/chinese.json` add dde-shutdown
*   feat: add check uid for lannch app and run cmd
*   fix: root window cursor is not `left_ptr` when use kwin as wm

[3.13.0] 2019-03-15
*   fix: panic if getSavedBrightnessTable return nil map
*   feat: run kwin as wm if install package dde-kwin
*   feat: support monitor-dependent scaling
*   fix(display): connect an external monitor, give incorrect brightness value
*   feat: play login sound only when login automatically
*   change(api): session manager add method RunCommandWithOptions

[3.12.2] 2019-03-01
*   feat(watchdog): SetLocked reset dde-lock task
*   chore(deb): remove dde-go-dbus-factory from build depends
*   chore: use go-dbus-factory
*   chore: use github.com/linuxdeepin/go-gir

[3.12.1] 2019-02-25
*   fix(display): monitors config lost

[3.12.0] 2019-02-22
*   feat: app proxy support google chrome
*   chore: set LANG and LANGUAGE for dbus daemon
*   feat(watchdog): add task for dde-lock
*   feat: limit ~/.dde-env only for deepin session
*   fix: can not play shutdown music

[3.11.0] 2019-01-25
*   feat: do not save env in .pam_environment
*   feat: no use sys config allow switch

## [3.10.0] - 2019-01-10
*   fix: modify the recommended scale factor calc

## [3.9.0] - 2019-01-09
*   fix: if the speaker is muted, still play the logout music

## [3.8.0] - 2018-12-28
*   feat(wm): Switcher add method AllowSwitch

## [3.7.0] - 2018-12-25
*   fix: Qt double click not working well on hiDPI devices
*   feat: CanSuspend check file `/sys/power/mem_sleep` existence
*   chore: compile with sw arch no longer needs to use gccgo

## [3.6.0] - 2018-12-10
*   fix: improve double clicks on touchscreen for Qt-based applications

## [3.5.0] - 2018-12-07
*   feat(fix-xauthority-perm): create .Xauthority if not exist

## [3.4.0] - 2018-11-23
*   feat: modify the meaning of allow switch
*   fix(xsettings): gtk window cursor size wrong

## [3.3.0] - 2018-11-09
*   fix(display): the disconnected output still on
*   feat(display): add touchscreen rotation supported

## [3.2.0] - 2018-10-26
*   fix(display): brightness setter auto is not good
*   fix(display): no listen xrandr event
*   chore: update changelog
*   fix: typo in `auto_launch/chinese.json`
*   chore: update changelog
*   chore: do not call gtk init
*   chore: `auto_launch/chinese.json` remove dde-control-center
*   chore: update changelog
*   fix: add booster-dtkwidget in auto launch chinese json
*   refactor: fix multiple typos
*   feat: run dde-welcome if deepin version changed
*   feat(wm): check wait for launcher wm

## [3.1.35.3] - 2018-09-07
*   fix: typo in `auto_launch/chinese.json`

## [3.1.35.2] - 2018-09-06
*   chore: do not call gtk init
*   chore: `auto_launch/chinese.json` remove dde-control-center
*   feat: run dde-welcome if deepin version changed

## [3.1.35.1] - 2018-08-31
*   fix: add booster-dtkwidget in auto launch chinese json
*   refactor: fix multiple typos

## [3.1.35] - 2018-07-31
*   fix(display): auto set scale factor to 2 in virtualbox
*   feat: allow session daemon run after dde-session-daemon part2 started

## [3.1.34] - 2018-07-23
*   chore(debian): update build depends
*   fix: test failed in the pbuilder environment
*   feat: auto set scale factor
*   chore(display): rename setGammaSize to setOutputCrtcGamma
*   chore(display): err message include stderr
*   feat: support dde-session-daemon 2 step start
*   feat: launch dde-welcome by exec
*   feat: setup environment in startdde
*   fix(display): set brightness error typo
*   chore: no import lib xgb/proto
*   chore: use go-x11-client
*   chore: do not auto launch dde-file-manager on x86 arch
*   perf(swapsched): use less cpu when swapsched is not really enabled

## [3.1.33] - 2018-06-13
*   feat(display): set default brightness of output to 1

## [3.1.32] - 2018-06-07
*   fix(wm): select wm in dde-wm-chooser but it doesn't work
*   feat: add bin fix-xauthority-perm
*   chore: update build-depends

## [3.1.31] - 2018-05-30
*   feat(wm): show osd after receive wm StartupReady signal
*   fix: can't log into session because of .Xauthority
*   feat(swapsched): check cgexec existence
*   chore: update makefile
*   chore: update makefile for arch `sw_64`

## [3.1.30] - 2018-05-23
*   fix(keyring): `check_login` infinite loop
*   fix(swapsched): data race problem

## [3.1.29] - 2018-05-14
*   fix(wm): fix genCardConfig when not found any video card
*   fix(watchdog): dde-polkit-agent running state check wrong
*   adjust `auto_launch.json`
*   fix: launch group
*   feat: wait wm launch
*   refactor: fix a typo
*   fix(debian): miss depends on libpam-gnome-keyring

main.go
/*
 * Copyright (C) 2014 ~ 2018 Deepin Technology Co., Ltd.
 *
 * Author:     jouyouyun <jouyouwen717@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package main

import "C"
import (
	"context"
	"flag"
	"fmt"
	"net"
	"os"
	"os/exec"
	"os/signal"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"

	dbus "github.com/godbus/dbus"
	accounts "github.com/linuxdeepin/go-dbus-factory/com.deepin.daemon.accounts"
	notifications "github.com/linuxdeepin/go-dbus-factory/org.freedesktop.notifications"
	x "github.com/linuxdeepin/go-x11-client"
	"github.com/linuxdeepin/startdde/display"
	"github.com/linuxdeepin/startdde/iowait"
	"github.com/linuxdeepin/startdde/watchdog"
	wl_display "github.com/linuxdeepin/startdde/wl_display"
	"github.com/linuxdeepin/startdde/wm_kwin"
	"github.com/linuxdeepin/startdde/xsettings"
	"github.com/linuxdeepin/go-lib/dbusutil"
	"github.com/linuxdeepin/go-lib/gettext"
	"github.com/linuxdeepin/go-lib/gsettings"
	"github.com/linuxdeepin/go-lib/log"
	"github.com/linuxdeepin/go-lib/proxy"
)

var logger = log.NewLogger("startdde")

var _options struct {
	noXSessionScripts bool
}

var _gSettingsConfig *GSettingsConfig

var globalCgExecBin string

var globalXSManager *xsettings.XSManager

var _xConn *x.Conn

var _useWayland bool

var _inVM bool

var _useKWin bool

var _homeDir string

func init() {
	flag.BoolVar(&_options.noXSessionScripts, "no-xsession-scripts", false, "")
}

func reapZombies() {
	// We must reap children process even we hasn't create anyone at this moment,
	// Because the startdde may be launched by exec syscall
	// in another existed process, like /usr/sbin/lighdm-session does.
	// NOTE: Don't use signal.Ignore(syscall.SIGCHILD), otherwise os/exec wouldn't work properly.
	//       And simply ignore SIGCHILD hasn't any helpful in here.
	for {
		pid, err := syscall.Wait4(-1, nil, syscall.WNOHANG, nil)
		if err != nil || pid == 0 {
			break
		}
	}
}

func shouldUseDDEKWin() bool {
	_, err := os.Stat("/usr/bin/kwin_no_scale")
	return err == nil
}

const (
	cmdKWin                = "/usr/bin/kwin_no_scale"
	cmdDdeSessionDaemon    = "/usr/lib/deepin-daemon/dde-session-daemon"
	cmdDdeDock             = "/usr/bin/dde-dock"
	cmdDdeDesktop          = "/usr/bin/dde-desktop"
	cmdLoginReminderHelper = "/usr/libexec/deepin/login-reminder-helper"
	cmdDdeHintsDialog      = "/usr/bin/dde-hints-dialog"

	loginReminderTimeout    = 5 * time.Second
	loginReminderTimeFormat = "2006-01-02 15:04:05"
	secondsPerDay           = 60 * 60 * 24
	accountUserPath         = "/com/deepin/daemon/Accounts/User"
)

func launchCoreComponents(sm *SessionManager) {
	setupEnvironments1()

	wmChooserLaunched := false
	if !_useWayland && _inVM {
		wmChooserLaunched = maybeLaunchWMChooser()
	}

	const waitDelayDuration = 7 * time.Second

	var wg sync.WaitGroup
	launch := func(program string, args []string, name string, wait bool, endFn func()) {
		if wait {
			wg.Add(1)
			sm.launchWaitCore(name, program, args, waitDelayDuration, func(launchOk bool) {
				if endFn != nil {
					endFn()
				}
				wg.Done()
			})
			return
		}

		sm.launchWithoutWait(program, args...)
	}

	coreStartTime := time.Now()
	// launch window manager
	if !_useWayland {
		_useKWin = shouldUseDDEKWin()
		if _useKWin {
			if wmChooserLaunched {
				wm_kwin.SyncWmChooserChoice()
			}
			launch(cmdKWin, nil, "kwin", true, func() {
				// 等待 kwin 就绪，然后让 dock 显示
				handleKWinReady(sm)
			})
		} else {
			wmCmd := _gSettingsConfig.wmCmd
			if wmCmd == "" {
				wmCmd = "x-window-manager"
			}
			launch("env", []string{"GDK_SCALE=1", wmCmd}, "wm", false, nil)
		}
	} else {
		var handleId dbusutil.SignalHandlerId
		var needDel = true
		handleId, err := sm.dbusDaemon.ConnectNameOwnerChanged(func(name, newOwner, oldOwner string) {
			if name == "org.kde.KWin" && newOwner != "" && oldOwner == "" {
				handleKWinReady(sm)
				sm.dbusDaemon.RemoveHandler(handleId)
			}
		})
		if err != nil {
			logger.Warningf("connect to signal err: %s, run handle after timeout", err)
			time.AfterFunc(time.Second*5, func() {
				handleKWinReady(sm)
			})
			needDel = false
		}

		hasOwner, err := sm.dbusDaemon.NameHasOwner(0, "org.kde.KWin")
		if err != nil {
			logger.Warning(err)
			return
		}

		if hasOwner {
			handleKWinReady(sm)
			if needDel {
				sm.dbusDaemon.RemoveHandler(handleId)
			}
		}
	}

	// 先启动 dde-session-daemon，再启动 dde-dock
	launch(cmdDdeSessionDaemon, nil, "dde-session-daemon", true, func() {
		var dockArgs []string
		if _useKWin {
			dockArgs = []string{"-r"}
		}
		launch(cmdDdeDock, dockArgs, "dde-dock", _useKWin, nil)
	})
	launch(cmdDdeDesktop, nil, "dde-desktop", true, nil)

	wg.Wait()
	logger.Info("core components cost:", time.Since(coreStartTime))
}

func handleKWinReady(sm *SessionManager) {
	sessionBus := sm.service.Conn()

	const dockServiceName = "com.deepin.dde.Dock"
	callDockShow := func() {
		logInfoAfter("call com.deepin.dde.Dock callShow")
		dockObj := sessionBus.Object(dockServiceName, "/com/deepin/dde/Dock")
		ctx, cancelFn := context.WithTimeout(context.Background(), 10*time.Second)
		defer cancelFn()
		err := dockObj.CallWithContext(ctx, dockServiceName+".callShow", dbus.FlagNoAutoStart).Err
		if err != nil {
			logger.Warning("call dde-dock callShow failed:", err)
		}
	}

	var sigHandleId dbusutil.SignalHandlerId
	sigHandleId, err := sm.dbusDaemon.ConnectNameOwnerChanged(func(name string, oldOwner string, newOwner string) {
		if name == dockServiceName && oldOwner == "" && newOwner != "" {
			callDockShow()
			sm.dbusDaemon.RemoveHandler(sigHandleId)
		}
	})
	has, err := sm.dbusDaemon.NameHasOwner(0, dockServiceName)
	if err != nil {
		logger.Warning(err)
	} else if has {
		callDockShow()
		sm.dbusDaemon.RemoveHandler(sigHandleId)
	}
}

var _mainBeginTime time.Time

func logDebugAfter(msg string) {
	elapsed := time.Since(_mainBeginTime)
	logger.Debugf("after %s, %s", elapsed, msg)
}

func logInfoAfter(msg string) {
	elapsed := time.Since(_mainBeginTime)
	logger.Infof("after %s, %s", elapsed, msg)
}

func greeterDisplayMain() {
	display.SetGreeterMode(true)
	// init x conn
	xConn, err := x.NewConn()
	if err != nil {
		logger.Warning(err)
		os.Exit(1)
	}
	// TODO
	display.Init(xConn, false)
	logger.Debug("greeter mode")
	service, err := dbusutil.NewSessionService()
	if err != nil {
		logger.Warning(err)
	}
	err = display.Start(service)
	if err != nil {
		logger.Warning(err)
	}
	err = display.StartPart2()
	if err != nil {
		logger.Warning(err)
	}
	service.Wait()
}

func main() {
	flag.Parse()
	if len(os.Args) > 0 && strings.HasPrefix(filepath.Base(os.Args[0]), "greeter") {
		// os.Args[0] 应该一般是 greeter-display-daemon
		greeterDisplayMain()
		return
	}

	initGSettingsConfig()

	_mainBeginTime = time.Now()

	gettext.InitI18n()
	gettext.BindTextdomainCodeset("startdde", "UTF-8")
	gettext.Textdomain("startdde")

	reapZombies()
	// init x conn
	xConn, err := x.NewConn()
	if err != nil {
		logger.Warning(err)
		os.Exit(1)
	}
	_xConn = xConn
	if _options.noXSessionScripts {
		runXSessionScriptsFaster(xConn)
	}
	_inVM, err = isInVM()
	if err != nil {
		logger.Warning("detect VM failed:", err)
	}
	var recommendedScaleFactor float64
	if os.Getenv("WAYLAND_DISPLAY") != "" {
		logger.Info("in wayland mode")
		_useWayland = true
	}
	display.Init(xConn, _useWayland)
	// TODO
	recommendedScaleFactor = display.GetRecommendedScaleFactor()

	service, err := dbusutil.NewSessionService()
	if err != nil {
		logger.Warning(err)
		os.Exit(1)
	}

	xsManager, err := xsettings.Start(xConn, recommendedScaleFactor, service, &display.ScaleFactorsHelper)
	if err != nil {
		logger.Warning(err)
	} else {
		globalXSManager = xsManager
	}

	sessionManager := newSessionManager(service)
	err = service.Export(sessionManagerPath, sessionManager)
	if err != nil {
		logger.Warning("export session sessionManager failed:", err)
	}
	err = service.RequestName(sessionManagerServiceName)
	if err != nil {
		logger.Warningf("request name %q failed: %v", sessionManagerServiceName, err)
	}
	logDebugAfter("before launchCoreComponents")

	err = display.Start(service)
	if err != nil {
		logger.Warning("start display part1 failed:", err)
	}

	launchCoreComponents(sessionManager)

	// 启动 display 模块的后一部分
	go func() {
		err := display.StartPart2()
		if err != nil {
			logger.Warning("start display part2 failed:", err)
		}
	}()

	go func() {
		initSoundThemePlayer()
		playLoginSound()
	}()

	err = gsettings.StartMonitor()
	if err != nil {
		logger.Warning("gsettings start monitor failed:", err)
	}
	proxy.SetupProxy()

	if _inVM {
		time.AfterFunc(10*time.Second, func() {
			logger.Debug("try to correct vm resolution")
			correctVMResolution()
		})
	}

	sysBus, err := dbus.SystemBus()
	if err != nil {
		logger.Warning(err)
		os.Exit(1)
	}
	sysSignalLoop := dbusutil.NewSignalLoop(sysBus, 10)
	sysSignalLoop.Start()

	sessionManager.start(xConn, sysSignalLoop, service)
	watchdog.Start(sessionManager.getLocked, _useKWin)

	if _gSettingsConfig.iowaitEnabled {
		go iowait.Start(logger)
	} else {
		logger.Info("iowait disabled")
	}

	go handleOSSignal(sessionManager)

	loginReminder()

	service.Wait()
}

func handleOSSignal(m *SessionManager) {
	var sigChan = make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGABRT, syscall.SIGTERM, syscall.SIGSEGV)

loop:
	for sig := range sigChan {
		switch sig {
		case syscall.SIGINT, syscall.SIGABRT, syscall.SIGTERM, syscall.SIGSEGV:
			logger.Info("received signal:", sig)
			break loop
		}
	}

	logger.Info("received unexpected signal, force logout")
	m.doLogout(true)
}

func doSetLogLevel(level log.Priority) {
	logger.SetLogLevel(level)
	if !_useWayland {
		display.SetLogLevel(level)
	} else {
		wl_display.SetLogLevel(level)
	}
	watchdog.SetLogLevel(level)
}

func loginReminder() {
	if !_gSettingsConfig.loginReminder {
		return
	}

	sysBus, _ := dbus.SystemBus()

	uid := os.Getuid()
	userPath := accountUserPath + strconv.Itoa(uid)

	user, err := accounts.NewUser(sysBus, dbus.ObjectPath(userPath))
	if err != nil {
		logger.Warning("failed to get user:", err)
	}

	res, err := user.GetReminderInfo(0)
	if err != nil {
		logger.Warning("failed to get reminder info:", err)
	}

	currentLoginTime, err := time.Parse("2006-01-02 15:04:05.999999999 -0700 MST", res.CurrentLogin.Time)
	if err != nil {
		logger.Warning("wrong time format:", err)
	}
	currentLoginTimeStr := currentLoginTime.Format(loginReminderTimeFormat)

	address := res.CurrentLogin.Address
	if address == "0.0.0.0" {
		// 若 IP 是「0.0.0.0」，获取当前设备的 IP
		address = getFirstIPAddress()
	}

	body := fmt.Sprintf("%s %s %s", res.Username, address, currentLoginTimeStr)

	// pam_unix/passverify.c
	curDays := int(time.Now().Unix() / secondsPerDay)
	daysLeft := res.Spent.LastChange + res.Spent.Max - curDays
	if res.Spent.Max != -1 && res.Spent.Warn != -1 {
		if res.Spent.Warn > daysLeft {
			body += " " + fmt.Sprintf(gettext.Tr("Your password will expire in %d days"), daysLeft)
		}
	}

	body += "\n" + fmt.Sprintf(gettext.Tr("%d login failures since the last successful login"), res.FailCountSinceLastLogin)

	bus, _ := dbus.SessionBus()
	notifi := notifications.NewNotifications(bus)
	sigLoop := dbusutil.NewSignalLoop(bus, 10)
	sigLoop.Start()
	notifi.InitSignalExt(sigLoop, true)

	// TODO: icon
	title := gettext.Tr("Login Reminder")
	actions := []string{"details", gettext.Tr("Details")}
	notifyId, err := notifi.Notify(0, "dde-control-center", 0, "preferences-system", title, body, actions, nil, 0)
	if err != nil {
		logger.Warningf("failed to send notify: %s", err)
		return
	}

	_, err = notifi.ConnectActionInvoked(func(id uint32, actionKey string) {
		if id != notifyId || actionKey != "details" {
			return
		}

		lastLoginTime, err := time.Parse("2006-01-02 15:04:05.999999999 -0700 MST", res.LastLogin.Time)
		if err != nil {
			logger.Warning("wrong time format:", err)
		}
		lastLoginTimeStr := lastLoginTime.Format(loginReminderTimeFormat)

		content := fmt.Sprintf("<p>%s</p>", res.Username)
		content += fmt.Sprintf("<p>%s</p>", address)
		content += "<p>" + fmt.Sprintf(gettext.Tr("Login time: %s"), currentLoginTimeStr) + "</p>"
		content += "<p>" + fmt.Sprintf(gettext.Tr("Last login: %s"), lastLoginTimeStr) + "</p>"
		content += "<p><b>" + fmt.Sprintf(gettext.Tr("Your password will expire in %d days"), daysLeft) + "</b></p>"
		content += "<br>"
		content += "<p>" + fmt.Sprintf(gettext.Tr("%d login failures since the last successful login"), res.FailCountSinceLastLogin) + "</p>"

		cmd := exec.Command(cmdDdeHintsDialog, title, content)
		err = cmd.Start()
		if err != nil {
			logger.Warning("failed to start dde-hints-dialog:", err)
			return
		}

		go func() {
			err = cmd.Wait()
			if err != nil {
				logger.Warning("failed to run dde-hints-dialog", err)
				return
			}
		}()
	})
	if err != nil {
		logger.Warning("connect to ActionInvoked failed:", err)
	}

	// 通知不显示在通知中心面板，故在时间到了后，关闭通知
	time.AfterFunc(loginReminderTimeout, func() {
		notifi.CloseNotification(0, notifyId)

		notifi.RemoveAllHandlers()

		sigLoop.Stop()
	})
}

func getFirstIPAddress() string {
	ifaces, err := net.Interfaces()
	if err != nil {
		panic(err)
	}

	for _, iface := range ifaces {
		if iface.Name == "lo" {
			continue
		}

		addrs, err := iface.Addrs()
		if err != nil {
			logger.Warningf("failed to get address of %s: %s", iface.Name, err)
			continue
		}

		for _, addr := range addrs {
			// remove the netmask
			return strings.Split(addr.String(), "/")[0]
		}
	}

	return "127.0.0.1"
}
